Rust: Add recipe for 1.21.0

nielx opened the pull request:
This recipe is built of binary artifacts. The main reason is that the 32
bit version can no longer be built on Haiku, because it crosses a hard
memory limit. Perhaps this changes later on. In any case, an added
advantage is that his makes the package buildable by the buildbots.

The line for SOURCES_3 is commented out. In here the sources that the
built is based on would be downloaded, mainly for the completeness of
the source packages. This gives an ascii encoding error in haikuporter.

nielx commented:
@waddlesplash @fbrosson please review?

fbrosson commented:
I've been thinking about something like:

```bash
REVISION="1"

case "$effectiveTargetArchitecture" in
	x86)
SOURCE_URI="http://dl.rust-on-haiku.com/dist/$portVersion/rust-$portVersion-i686-unknown-haiku.tar.xz"
CHECKSUM_SHA256="465db013876ead01686c991f1fa63ef4a6687d3cabbcde81f5a756e367989cdf"
		;;
	x86_64)
SOURCE_URI="http://dl.rust-on-haiku.com/dist/$portVersion/rust-$portVersion-x86_64-unknown-haiku.tar.xz"
CHECKSUM_SHA256="cbb7f705398b76c5e2dd18119cdb9af2c042720449479b482a39309fb270ded8"
		;;
	*)
SOURCE_URI="http://dl.rust-on-haiku.com/dist/rustc-$portVersion-src.tar.xz"
CHECKSUM_SHA256="7741c2f3652813b7ecd0233305a5ca0de561fea2849049651a79521d2c437f0f"
		;;
esac

ARCHITECTURES="!x86_gcc2 ?x86 x86_64"
SECONDARY_ARCHITECTURES="x86"

cargoVersion="0.22.0"
rlsVersion="0.121.0"
```

I wanted to try this but did not have enough time. (Of course, there would be other changes too, in `BUILD` and `INSTALL`.)
So this is just a draft, completely unfinished and untested...

BTW, I also don't know if it would be acceptable to have different `SOURCE_DIR`s based on `$effectiveTargetArchitecture`.
Well, I cannot think of any reason why it would not be OK, except, maybe, if we want the source packages to be independent of the architecture where the source package is being "built".

nielx commented:
The conditional downloading could be an option. 

An alternative could be that the actual source is used as the 'source', and that in the build step we download the binary artifacts for the actual platform. That will generate proper source packages.

fbrosson commented:
Yeah, there might be something cool to do here, but I am missing some bits.
In particular, I don't know if it is possible to insert a test that would allow the recipe to figure out what kind of package haikuporter is creating, I mean a binary (for one of the architectures, including `any`) or a source package.

OK, I see what you mean. We would define only one `SOURCE_URI`, and it would be that of the real sources. And in `BUILD` we would download the tarball with the binary artifacts that match the target arch.
Well, I don't know if this is would be OK because my understanding is that all the tarballs should be downloaded by haikuporter itself, before any of the `PATCH`, `BUILD` or `INSTALL` functions are called.
But maybe we could accept an exception for this recipe.
Let's see what the other contributors think about downlading the binary artifacts in `BUILD`.

waddlesplash commented:
Typically packages that only download binary artifacts are prefixed with `_bin`. Otherwise, fine by me.

nielx commented:
@waddlesplash: would you suggest keeping the recipe as is? Or would we get the binary artifacts in the build stage?

About the _bin suffix... the main issue is that it is impossible to build Rust on a 32-bit system because it runs out of memory. It is possible to build it on 64 bit systems (though I am also running into memory issues with rust 1.27, but that's an issue for later). 

Would you suggest that I post both the rust-1.21.0 and rust_bin-1.21.0 recipes, where the source recipe will only work on 64 bit systems?

waddlesplash commented:
> Would you suggest that I post both the rust-1.21.0 and rust_bin-1.21.0 recipes, where the source recipe will only work on 64 bit systems?

Yep, that sounds best.

alaviss commented:
I'm against downloading **anything** during `BUILD`, as:

- Proxies are not respected. HaikuPorter's chroot doesn't inherit environment variables from it's caller, so proxy settings doesn't apply there.
- The results are not cached.

waddlesplash commented:
Oh, to clarify: a `_bin` recipe downloads its files as part of SOURCE_URL, and then would set `NO_SOURCE_PACKAGES`. So in that case nothing would be downloaded during `build`.

nielx commented:
Are you sure the _bin package should not generate a source package? Because it does...

Furthermore, it no longer replaces earlier versions of the package...

waddlesplash commented:
Sorry, I think it's DISABLE_SOURCE_PACKAGES.

For replacing, use REPLACES. There are some other recipes that already do.

nielx commented:
I would argue against renaming the recipe to rust_bin, since it is very unlikely it conflicts with a source package in the future (rust builds seem to get heavier instead of lighter). Furthermore, it would complicate a future where we would split up the packages into rust(_runtime), rust_devel, rust_rustfmt.

waddlesplash commented:
> since it is very unlikely it conflicts with a source package in the future (rust builds seem to get heavier instead of lighter). 

That's not really the point; the point is if HaikuPorter does not build it but just archives binary artifacts, the recipe is called `_bin`.

> Furthermore, it would complicate a future where we would split up the packages into rust(_runtime), rust_devel, rust_rustfmt.

How so? In those cases, we would do `packageEntries runtime` for the Rust runtime package in both the regular and `_bin` recipes, and add the appropriate provides. A few extra lines is all.

waddlesplash commented:
This is to fix https://dev.haiku-os.org/ticket/14233 though. So not sure that's a problem here?

nielx commented:
Updated the recipe again, now with REPLACE section and DISABLE_SOURCE_PACKAGE

fbrosson commented:
There's something that is not as nice as I think it could be. It's the beginning of `INSTALL()`:
```
	# we will manually invoke the install scripts
	if [ $effectiveTargetArchitecture = x86 ]; then
		cd $sourceDir
	fi
	if [ $effectiveTargetArchitecture = x86_64 ]; then
		cd $sourceDir2/rust-1.21.0-x86_64-unknown-haiku
	fi
```
I think we could avoid this by using conditional definitions for `SOURCE_URI`, as suggested in [my first comment](https://github.com/haikuports/haikuports/pull/2647#issuecomment-399678783).

That said, I'm OK with the merge. It's just that I haven't tested the recipe myself and that I'm not very fond of the above bloc, so I'd rather let someone else do the merge :)

fbrosson commented:
@nielx: That's great!
(I did not test it, but I'm OK with assuming it works, because I guess you did check the variant.)

@waddlesplash: are you OK with the new variant?
I might try the recipe and then merge the PR if everything is OK, but please feel free to merge it when you wish, because I don't know how long it could take me to try the recipe.

@nielx, if you are 100% sure everything is OK (and if @waddlesplash is OK with the variant) then I can merge it without testing :)

nielx commented:
@fbrosson While you won't get a 100% guarantee, I have tested the build on x86 and x86_64, and both work fine.

waddlesplash commented:
Looks good.

fbrosson commented:
I've got a few tiny changes to suggest.

fbrosson commented on dev-lang/rust/rust_bin-1.21.0.recipe:
```diff
@@ -0,0 +1,94 @@
+rlsVersion="0.121.0"
+
+PROVIDES="
+	rust$secondaryArchSuffix = $portVersion
```
Hmm, I wonder if we should not change this to:
```diff
 PROVIDES="
-	rust$secondaryArchSuffix = $portVersion
+	rust${secondaryArchSuffix}_bin = $portVersion
```
and add a
```
CONFLICTS="
	rust$secondaryArchSuffix
	"
```

I think this was discussed but I don't remember what was the conclusion, so I might very well be missing something...

I also wonder if we should not define a `commandBinDir` and `commandSuffix` to make the recipe `x86_64-hybrid`-_ready_.
(We would append a `commandSuffix` to `cmd:rust{c,doc,_lldb}`.)

nielx replied:
The outcome was that the recipe now defines a REPLACES line, which forces the uninstall of earlier non_bin recipes. About being 64 bit ready: I foresee some problems with hardcoded paths. I think it is better to look at the situation then and solve the problems then. Rust releases every six weeks, so there will be opportunity in the future.

I have made the smaller changes. Thanks for reviewing!

fbrosson replied:
I think the clock of your VM is not correctly set. I suspect that:
1) your real timezone is UTC+x (i.e. your latitude is close to 15*x East)
2) the clock of your VM is set to UTC and your localtime is being used as the UTC time.

This is not very important, but there is a little side effect: GitHub shows your commit updates below the last message, unless that message was added at least x hours after the last commit update.

For our visitors, the following commands can help finding the right settings:
```
date
TZ=UTC date
```

fbrosson commented:
Sorry, my previous message (about the timezone settings) might be wrong. Don't worry about it.

nielx commented:
First of all, I will test whether HaikuDepot can handle uninstalling it.

Furthermore I would really like to reconsider giving this the _bin suffix. @waddlesplash mentions that this is convention. I find some examples of it in the tree, for example the [free pascal](https://github.com/haikuports/haikuports/tree/ab6d8aa9760bf9cc64ec9d49fb4353a4f5af9eca/dev-lang/fpc) recipe, but here there is the situation that the 2.6.2 recipe requires the prebuilt binary as a build dependency. It is very unlikely that at this point there will ever be a parallel Rust 1.21.0 source and binary recipe, and I am wondering whether it is not better to opt into this convention when we will start providing two parallel recipes.

pulkomandy commented:
On Mon, Jul 02, 2018 at 12:11:32PM +0000, Niels Sascha Reedijk wrote:
> Furthermore I would really like to reconsider giving this the _bin suffix. @waddlesplash mentions that this is convention. I find some examples of it in the tree, for example the [free pascal](https://github.com/haikuports/haikuports/tree/ab6d8aa9760bf9cc64ec9d49fb4353a4f5af9eca/dev-lang/fpc) recipe, but here there is the situation that the 2.6.2 recipe requires the prebuilt binary as a build dependency. It is very unlikely that at this point there will ever be a parallel Rust 1.21.0 source and binary recipe, and I am wondering whether it is not better to opt into this convention when we will start providing two parallel recipes. 

I agree. We didn't do _bin for Wonderbrush or liblayout, for example. It
is only useful if the _bin recipe is eventually used to build a non _bin
one (ie solve bootstrapping problems). Since this won't be the case here
and no alternate, non _bin recipe will be provided, there is no need for
an _bin suffix.

Unless we plan to provide a recipe that attempts to build from sources,
even if it is currently not working because of memory constraints?

fbrosson commented:
That's funny :)
It turns out that the only `*_bin-*.recipe` we already have are the 3 fpc_bin recipes.
And they all have `fpc_bin` in their `PROVIDES`.
Dropping the `_bin` in the recipe's file name, i.e. making an exception to the conventions, might be acceptable, but now that I saw that the 3 fpc_bin recipes have the `_bin` suffix in their `PROVIDES`, I really think it is what we should do for rust.

(The only difference compared to fpc would be that rust would have a `CONFLICTS`, contrary to fpc, and that it would use `$secondaryArchSuffix`, while `fpc_bin-2.6.2.recipe` is missing the `$secondaryArchSuffix` in its `PROVIDES`.)

OK, so we have 2 scenari here:

1) rename `rust_bin-1.21.0.recipe` to `rust-1.21.0.recipe`.

2) fix the `PROVIDES`:
```diff
 PROVIDES="
-	rust$secondaryArchSuffix = $portVersion
+	rust_bin$secondaryArchSuffix = $portVersion
```
and add a
```
CONFLICTS="
	rust$secondaryArchSuffix
	"
```
I'm OK with both scenari, so if everyone is OK with the rename, that's probably the best choice.

That said, I might try to kick a build of `rust_bin_x86-1.21.0` in a VM with a lot of RAM, i.e. around 24 GB.
In case the build succeeds it could be worth keeping the `_bin` suffix so that we could use it to build a _non-bin_ rust using  _rust_bin_.

nielx commented:
> That said, I might try to kick a build of rust_bin_x86-1.21.0 in a VM with a lot of RAM, i.e. around 24 GB.

For that you would need a recipe that actually builds from source, which I haven't written yet (but I can).

Here's the thing, it could be useful to still have rust-recipes next to the rust_bin recipes, and like fpc it would be an added advantage because the non-bin recipes can use the bin recipes as bootstrap compiler. The question is why we would do this: the source recipe would only work on x86_64 with lots of RAM (will that work on the buildbot?) and it will make the namespace for the rust toolset even more complex. At the other hand, it is nice to have a build recipe that at least documents the steps to build Rust, even if there is always a dependency on a binary pre-installed package.

I'm torn on this.

fbrosson commented:
Well, I'd say that not having enough RAM in the builtbot right now should not matter, because this can be changed later. Also, users with a lot of RAM could provide feedback about their building rust non-bin.
So, since the changes for having rust_bin are easy to do (i.e. fix the `PROVIDES` and add a `CONFLICTS`, it might be worth following this scenario.

BTW, I see another advantage of having both rust and rust_bin: we would be able to:
1) build and install rust_bin.
2) build rust with rust_bin.
3) uninstall rust_bin and install rust.
4) build rust again, but this time using the rust we had previously built with rust_bin.
5) compare the two hpkgs we got when building rust. If they differ then we would know there might be something wrong in rust_bin.

> For that you would need a recipe that actually builds from source, which I haven't written yet (but I can).

Yep. BTW, I guess the recipe for rust would be quite close to the one for rust_bin. Hmm, if we add a recipe for rust non-bin in this PR then we can drop the 3 lines in the `case` ... `esac` that correspond to the `*)` case, since they would already be in rust non-bin.

pulkomandy commented:
On Mon, Jul 02, 2018 at 01:03:17PM +0000, Niels Sascha Reedijk wrote:
> At the other hand, it is nice to have a build recipe that at least documents the steps to build Rust, even if there is always a dependency on a binary pre-installed package.
> 
> I'm torn on this. 

Yes, if we can get a "documentation" recipe that would be great already.
We can just mark it as broken until we figure out a way to have it use
less RAM, but at least we know the steps needed and other people can
take a look.

fbrosson commented:
I've just built `rust_bin_x86-1.21.0` (using haikuporter's `--strict-policy` as I usually do) and it complains about missing provides for each of the libs:
```
Warning: POLICY ERROR: no matching provides "lib:*_x86" for "lib/*.so"
````
with X from each of the following list.
<Details>

```
libarena-080e5f9875b9c42b
libfmt_macros-6964d7d69f6440be
libgraphviz-d5e5f01e89ae6ac2
libproc_macro-d9a4e2bb7dd81fd8
librustc_allocator-de613be9eee6f071
librustc_back-40165c3e0982053f
librustc_borrowck-96ef2bb379fe200b
librustc_const_eval-cd9682353f50c131
librustc_const_math-992294e5e7c5658f
librustc_data_structures-5de99b114d61f058
librustc_driver-9fdcb28abc4a7695
librustc_errors-df5507fb5d85410c
librustc_incremental-33f9339249b6493a
librustc_lint-bc193f788b87a9f0
librustc_llvm-4c6de4cecf7c9e2b
librustc_metadata-ece6cab971540da8
librustc_mir-4537b141c0f73978
librustc_passes-9dcb488c533665ce
librustc_platform_intrinsics-22a04034150339f9
librustc_plugin-1e80ea04086d8d70
librustc_privacy-7fe7ad169961bdd3
librustc_resolve-a2e0067f2bd97460
librustc_save_analysis-0a21c3958d83188b
librustc_trans_utils-85c62182514c6ad9
librustc_trans-2b2a2cb163ccb259
librustc_typeck-71b85ff351a306b4
librustc-afc8ae00e810a630
libserialize-b90d15281e57cc35
libstd-9b31c1415ffb61d4
libsyntax_ext-5443cb4566e261f9
libsyntax_pos-5414099c8b646821
libsyntax-5ee0947850fd918d
libterm-72a275b478b0ccd8
libtest-fa4f781426809cbd
```
</Details>
BTW, the names of these libs include sequences of 16 hex digits:

nielx commented:
The rationale is that rustc and cargo by design resort to static linking, so there is never any external dependency on these libraries. I will leave it for now. Tonight I'll update the PR with a source recipe.

fbrosson commented on dev-lang/rust/rust_bin-1.21.0.recipe:
```diff
@@ -0,0 +1,94 @@
+	cmd:rustdoc = $portVersion
+	cmd:rust_gdb = $portVersion
+	cmd:rust_lldb = $portVersion
+	cmd:cargo$secondaryArchSuffix = $cargoVersion
```
It turns out that adding `commandBinDir` and `commandSuffix` works fine here.
Shall we use it?
And shall we tweak the `PROVIDES` to have rust_bin?
The changes for the above suggestions are:
```diff
 DISABLE_SOURCE_PACKAGE=yes
 
+commandSuffix=$secondaryArchSuffix
+commandBinDir=$binDir
+if [ "$targetArchitecture" = _x86_gcc2 ]; then
+	commandSuffix=
+	commandBinDir=$prefix/bin
+fi
+
 cargoVersion="0.22.0"
 rlsVersion="0.121.0"
 
 PROVIDES="
-	rust$secondaryArchSuffix = $portVersion
+	rust_bin$secondaryArchSuffix = $portVersion
-	cmd:rustc = $portVersion
-	cmd:rustdoc = $portVersion
-	cmd:rust_gdb = $portVersion
-	cmd:rust_lldb = $portVersion
+	cmd:rustc$commandSuffix = $portVersion
+	cmd:rustdoc$commandSuffix = $portVersion
+	cmd:rust_gdb$commandSuffix = $portVersion
+	cmd:rust_lldb$commandSuffix = $portVersion
-	cmd:cargo$secondaryArchSuffix = $cargoVersion
+	cmd:cargo$commandSuffix = $cargoVersion
-	cmd:rls = $rlsVersion
+	cmd:rls$commandSuffix = $rlsVersion
 	"
```
and later, in `INSTALL`:
```diff
 	./$module/install.sh                               \
 			--prefix=$prefix                           \
+			--bindir=$commandBinDir                    \
 			--docdir=$developDocDir                    \
```
```diff
-	# move the cargo and binaries (in case of a secondary arch)
-	if [ -n "$secondaryArchSuffix" ]; then
- 		mkdir -p $binDir
- 		mv $prefix/bin/cargo $binDir/cargo
- 	fi
```

nielx replied:
I don't quite understand what you are doing here, but whatever it is, it should comply with the following:
1. cargo depends on being able to call an executable called `rustc`. 
2. `rustc` will determine its sysroot relative to its own location. Specifically it looks for the `rustlib` directory at the `../lib/rustlib` location. That's why on x86_gcc2h only cargo has this suffix, not rustc. 

As per 2, the rationale that cargo has the suffix, is to remind the user to perform a setarch x86 before trying to compile rust projects, as not to cause problems with using the wrong compiler and linker.

fbrosson replied:
OK, I got it :)
(I'll update my comment asap.)

fbrosson replied:
In case it is better, for a x86 secondary arch build, to have all the runtimes in `/system/bin/x86/` then we should:
* keep your version of the _"move the cargo and binaries"_ bloc;
* ignore my suggestion to add definitions for `commandBinDir` and `commandSuffix`;
* replace all occurences of `$commandSuffix` by `$secondaryArchSuffix`.

nielx replied:
Actually it is better not to have binaries in /system/bin/x86, but in /system/bin.

fbrosson replied:
Great! 
My suggestions [above](https://github.com/haikuports/haikuports/pull/2647#discussion_r199521705) handles this nicely :)

BTW, I forgot to answer your question about why I'm suggesting to define `commandBinDir` and `commandSuffix`. It's to make sure that, on x86_64-hybrid, the `cmd:*` provided by `rust_bin_x86` won't shadow the `cmd:*` provided by `rust_bin`.
This will make sure recipes which might have e.g. `cmd:rustc` in their `BUILD_REQUIRES` will pull `rust_bin` and not `rust_bin_x86`.
(On x86_gcc2 there is no risk of the `cmd:*` for the secondary arch build shadowing those of the primary arch... because rust is not available on x856_gcc2 primary arch.)

Yeah, as I wrote in #2690, nobody will try to build `rust_x86` on `x86_64-hybrid`, but using a `commandSuffix` will allow to have a valid recipe in all cases and avoid a situation where two different builds of rust_bin (that is, for x64_64 primary arch and for x86 secondary arch) would provide the same `cmd:*`.

nielx commented:
I've updated the PR with both a rust_bin and a rust recipe. Please note that the rust recipe currently does not build, because it complains about not finding all the required packages. I don't have the skills to debug this.

The reason I added a rust_bin subdirectory in dev_lang, is because haikuporter otherwise wants to share the work directory.

fbrosson commented on dev-lang/rust/rust-1.21.0.recipe:
```diff
@@ -0,0 +1,211 @@
+DESCRIPTION="Rust is a systems programming language that runs blazingly fast, prevents almost all crashes*, and eliminates data races."
+HOMEPAGE="http://www.rust-lang.org/"
+LICENSE="MIT"
+COPYRIGHT="2017, The Rust Project Developers"
```
```diff
-DESCRIPTION="Rust is a systems programming language that runs blazingly fast, prevents almost all crashes*, and eliminates data races."
+DESCRIPTION="Rust is a systems programming language that runs blazingly fast, \
+prevents almost all crashes*, and eliminates data races."
-HOMEPAGE="http://www.rust-lang.org/"
+HOMEPAGE="https://www.rust-lang.org/"
+COPYRIGHT="2017 The Rust Project Developers"
 LICENSE="MIT"
-COPYRIGHT="2017, The Rust Project Developers"
+REVISION="1"
```

fbrosson commented on dev-lang/rust/rust-1.21.0.recipe:
```diff
@@ -0,0 +1,211 @@
+CHECKSUM_SHA256_12="82fa7ada1bdadbba9f780847dc4f9340f15e5f137d838c98ca7e30ec3dda6684"
+SOURCE_DIR="rust-$srcGitRev"
+PATCHES="rust-$portVersion.patchset"
+REVISION="1"
```
```diff
-REVISION="1"
+
```

fbrosson commented on dev-lang/rust/rust-1.21.0.recipe:
```diff
@@ -0,0 +1,211 @@
+	lib:libssl$secondaryArchSuffix
+	lib:libz$secondaryArchSuffix
+	"
+BUILD_REQUIRES="
```
```diff
 	"
+
 BUILD_REQUIRES="
```
(adding a blank line)

fbrosson commented on dev-lang/rust/rust-1.21.0.recipe:
```diff
@@ -0,0 +1,211 @@
+	rust_bin$secondaryArchSuffix = $portVersion
+	"
+CONFLICTS="
+	rust_bin$secondaryArchSuffix
```
Wouldn't it be nicer to remove the `CONFLICTS` in rust non-bin and add the following one to rust_bin?
```
CONFLICTS="
	rust$secondaryArchSuffix
	"
```

(BTW, the prefered place for the `CONFLICTS` definition is just after `REQUIRES`.)

BTW, if replacing the `rust_bin$secondaryArchSuffix = $cargoVersion` (in `BUILD_REQUIRES`) by `cmd:cargo$secondaryArchSuffix = $cargoVersion` works fine then it might be a good idea to do so, as this would allow to build rust with rust itself or with rust_bin.
Hmm, I might be missing something, so this is really just an untested suggestion :)

fbrosson commented on dev-lang/rust/rust-1.21.0.recipe:
```diff
@@ -0,0 +1,211 @@
+	cp -r $sourceDir2/compiler-rt-$srcGitRev2 src/libcompiler_builtins/compiler-rt
+
+	# write the build configuration
+	cat >config.toml <<- EOL
```
Suggestion:
```diff
-	cat >config.toml <<- EOL
+	tr -d '\t' >config.toml <<- EOL
```
(A variant would be to keep the `cat` command and remove each tab between lines 118 to 140.)

fbrosson commented on dev-lang/rust_bin/rust_bin-1.21.0.recipe:
```diff
@@ -0,0 +1,94 @@
+SUMMARY="Modern and safe systems programming language"
+DESCRIPTION="Rust is a systems programming language that runs blazingly fast, prevents almost all crashes*, and eliminates data races."
+HOMEPAGE="https://www.rust-lang.org/"
+COPYRIGHT="2017, The Rust Project Developers"
```
Suggestion :)
```diff
-COPYRIGHT="2017, The Rust Project Developers"
+COPYRIGHT="2017 The Rust Project Developers"
```

nielx commented:
@fbrosson  One last thing I just noticed, I see the rust_bin recipe has set the x86_64 and the secondary target x86 to untested, so the buildbot won't pick it up. Could you make a quick commit to change that?
